File: CPSC-3200/Programming Assignment 5/customer.cs
// Author: Clay Nguyen
// December 3, 2021
// Last Revision - December 3, 2021

// Class Invariant: customer object is instantiated with an unsigned balance
// and a unsigned account Number. Balance will change by adding money or buying
// items from a vendor
// 
// Interface Invariant: Client must inject dependencies into the constructor
// and the functions correctly. When buying, users only need to call a buy function with
// the correct parameters
//

using System;
namespace P5
{
    interface ICustomer
    {
        void addMoney(uint money);
        bool purchase(uint money);
        bool buyOne(string name, Vendor vendor);
        bool buy(Vendor vendor);
        uint getBalance();
    }

    public class Customer : ICustomer
    {
        private uint balance;
        private uint accountNumber;

        // Pre-Condition: Must inject a valid uint balance and accNum into the ctor
        // Post-Condition: Properties of the class are set here
        public Customer(uint bal, uint accNum)
        {
            balance = bal;
            accountNumber = accNum;
        }

        // Pre-Condition: Must inject a valid uint to be added to the balance
        // Post-Condition: Balance is updated to include the additional money added
        public void addMoney(uint money)
        {
            balance += money;
        }

        // Pre-Condition: Must inject a valid uint to be added to the balance
        // Post-Condition: if the balance is greater than the amount passed in,
        // then the money will be subtracted from the balance.
        public bool purchase(uint money)
        {
            if (money <= balance)
            {
                balance -= money;
                return true;
            }
            return false;
        }

        // Pre-Condition: Must inject a valid string name and Vendor object to buy items
        // Post-Condition: returns true if the item is purchased and the item is sold,
        // and not in the linked list of the vendor anymore
        public virtual bool buyOne(string name, Vendor vendor)
        {
            float price = vendor.getPrice(name);
            if (price == -1) { return false; }
            uint payment = (uint)price;
            bool bought = vendor.Sell(name);
            if (bought == true)
            {
                return purchase(payment);
            }
            return false;
        }


        // Pre-Condition: Must inject a valid vendor into the function
        // Post-Condition: Buys/Deletes items from the vendor of choice if the customer
        // has enough money. 
        public virtual bool buy(Vendor vendor)
        {
            vendor.CleanStock();
            string[] entreeNames = vendor.randomEntree();
            string entree1 = "";
            string entree2 = "";
            string entree3 = "";
            if (entreeNames.Length > 3)
            {
                entree1 = entreeNames[0];
                entree2 = entreeNames[1];
                entree3 = entreeNames[2];
            }
            else
            {
                return false;
            }
            float entree1price = vendor.getPrice(entree1);
            float entree2price = vendor.getPrice(entree2);
            float entree3price = vendor.getPrice(entree3);

            if (getBalance() >= (entree1price + entree2price + entree3price))
            {
                buyOne(entree1, vendor);
                buyOne(entree2, vendor);
                buyOne(entree3, vendor);
                return true;
            }
            else if (getBalance() >= (entree1price + entree2price))
            {
                buyOne(entree1, vendor);
                buyOne(entree2, vendor);
                return true;
            }
            else if (getBalance() >= (entree2price + entree3price))
            {
                buyOne(entree2, vendor);
                buyOne(entree3, vendor);
                return true;
            }
            else if (getBalance() >= (entree1price + entree3price))
            {
                buyOne(entree1, vendor);
                buyOne(entree3, vendor);
                return true;
            }
            else
            {
                // cannot buy for whatever reason
                return false;
            }
        }

        // Pre-Condition: None
        // Post-Condition: returns the balance of the customer
        public uint getBalance()
        {
            return balance;
        }
    }
}

// Implementation Invariant: implemented balance as an unsigned int to never
// allow balances to be negative. implemented the encapsulation of vendor type
// checking and customer balance checking in the buy functions

File: Object-Oriented-Development/Programming Assignment 2/customer.cpp
//AUTHOR: Clay Nguyen
//DATE: October 22, 2021
//REVISION HISTORY: FIRST SUBMISSION

#include <iostream>
#include "customer.h"
#include "vendor.h"


Customer::Customer(unsigned int bal, unsigned int accNum) : balance(bal), accountNumber(accNum) {
}

void Customer::addMoney(unsigned int money) {
    balance += money;
}

bool Customer::purchase(unsigned money){
  if(money <= balance){
    balance -= money;
    return true;
  }
  return false;
}

bool Customer::buy(Vendor& vendor) {
  vendor.CleanStock();
  string entreeName;
  string entree2;
  string entree3;
  Entree firstEntree;
  Entree secondEntree;
  Entree thirdEntree;

  
  for (int i = 0; i < 3; i++){
    if(vendor.canBuy() == true){
      entreeName = vendor.getTop();
      if(vendor.getPrice(entreeName) <= GetBalance()){
        if (i == 0){
          firstEntree = buyOne(entreeName, vendor);
        } else if (i == 1){
          secondEntree = buyOne(entreeName, vendor);
        } else {
          thirdEntree = buyOne(entreeName, vendor);
        }
      }
    }
  }
  if (firstEntree.GetName() != "" && secondEntree.GetName() != "" && thirdEntree.GetName() != "")
    {
      Entree returnEntree(firstEntree);
      returnEntree += secondEntree;
      returnEntree += thirdEntree;
      cout << "Bought three entrees" << endl << endl;
      cout << returnEntree << endl << endl;
      return true;
    }
  else if (firstEntree.GetName() != "" && secondEntree.GetName() != "")
    {
      Entree returnEntree(firstEntree);
      returnEntree += secondEntree;
      cout << "Bought two entrees" << endl << endl;
      cout << returnEntree << endl << endl;
      return true;
    }
  else if (secondEntree.GetName() != "" && thirdEntree.GetName() != "")
    {
      Entree returnEntree(secondEntree);
      returnEntree += thirdEntree;
      cout << "Bought two entrees" << endl << endl;
      cout << returnEntree << endl << endl;
      return true;
    }
  else if (firstEntree.GetName() != "" && thirdEntree.GetName() != "")
    {
      Entree returnEntree(firstEntree);
      returnEntree += thirdEntree;
      cout << "Bought two entrees" << endl << endl;
      cout << returnEntree << endl << endl;
      return true;
    }
  else
    {
      // cannot buy for whatever reason                                                           
      return false;
    }
  return false; 
  
}

Entree Customer::buyOne(string name, Vendor& vendor) {
  Entree temp;
  vendor.CleanStock();
  float price = vendor.getPrice(name);
  if(price != -1) {
    unsigned payment = (unsigned)price;
    bool bought = purchase(payment);
    if(bought)
      {
        temp = vendor.Sell();
      }
  }
  return temp;

}

unsigned int Customer::GetBalance() {
    return balance;
}

Customer Customer::operator+(const Customer& b){
  // make and return local copy. add the balances 
  Customer local(b);
  local.balance = balance + b.balance;
  local.accountNumber = accountNumber + b.accountNumber;
  return local;
}

Customer Customer::operator-(const Customer& b){
  // make and return local copy. add the balances
  Customer local(b);
  if(balance < b.balance){
    return local;
  } else {
    local.balance = balance - b.balance;
  }
  return local;
}


Customer Customer::operator+(int inc){
  Customer local(balance, accountNumber);
  local.balance = local.balance + inc;
  return local;
}

Customer Customer::operator-(int inc){
  Customer local(balance, accountNumber);
  if(balance < inc){
    return local;
  } else {
    local.balance = local.balance - inc;
  }
  return local;
}

Customer operator+(int num, Customer& b){
  return b + num;
}

Customer operator-(int num, Customer& b){
  return b - num;
}


Customer Customer::addOne(){
  ++balance;
  return *this;
}

Customer Customer::subOne(){
  --balance;
  return *this;
}

Customer Customer::operator++(){
  return addOne();
}

Customer Customer::operator--(){
  return subOne();
}

Customer Customer::operator++(int x){
  Customer oldState = *this;
  addOne();
  return oldState;
}

Customer Customer::operator--(int x){
  Customer oldState = *this;
  subOne();
  return oldState;
}

bool Customer::operator<(const Customer& b){
  if (balance < b.balance){
    return true;
  } else {
    return false;
  }
}

bool Customer::operator<=(const Customer& b){
  if (balance <= b.balance){
    return true;
  } else {
    return false;
  }
}

bool Customer::operator==(const Customer& b){
  if (balance == b.balance){
    return true;
  } else {
    return false;
  }
}

bool Customer::operator!=(const Customer& b){
  if (balance != b.balance){
    return true;
  } else {
    return false;
  }
}

bool Customer::operator>(const Customer& b){
  if (balance >= b.balance){
    return true;
  } else {
    return false;
  }
}

bool Customer::operator>=(const Customer& b){
  if (balance >= b.balance){
    return true;
  } else {
    return false;
  }
}

// in reality, should not have assignment operator, would lead to
// two people having the same balance and account Number
Customer& Customer::operator=(const Customer& b){
  return *this;
}

Customer& Customer::operator+=(const Customer& b){
  balance += b.balance;
  return *this;
}

Customer& Customer::operator-=(const Customer& b){
  if (balance < b.balance){
    return *this;
  }
  balance += b.balance;
  return *this;
}

ostream& operator<<(ostream& out, const Customer& customer){
  out << "Customer Balance: " << customer.balance << endl;
  return out;
}
